### PR TITLE
Allow correct filtering even if multiple servers share the same group names. Plus: README.md image upload GD lib. Fixes #72 and Fixes #74

### LINKED ISSUE #72: Multiple servers with same "include filter" group names do not work [PR exists]
TL;DR: I will provide a pull request for this issue.

Steps to reproduce:
* Configure multiple (different servers)
* All serves have a group defined. All with the same name "2FritzBox" and each group has some vCards of the respective server inside
* configure an "include" filter like so:
```
    'filters' => [
        'include' => [
            'group' => [
                '2FritzBox'
            ],
        ]
    ],
``` 
* Asume server#1 has 200 vCards with 100 vCards in the group "2FritzBox" 
   and server#2 has 100 vCards with 50 vCards in the "2FritzBox" group.

Expected: after filtering 150 cards should be imported to FritzBox
But actually only 50 cards survive the filter.

### LINKED ISSUE #74: Image upload not working (missing installation step) [PR exists]
A `run` command with parameter `-i`terminates with the following error:
```
PHP Fatal error:  Uncaught Error: Call to undefined function imagecreatefromstring() in /home/pi/bin/carddav2fb/src/functions.php:59
Stack trace:
#0 /home/pi/bin/carddav2fb/src/RunCommand.php(64): Andig\uploadImages(Array, Array)
#1 /home/pi/bin/carddav2fb/vendor/symfony/console/Command/Command.php(255): Andig\RunCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#2 /home/pi/bin/carddav2fb/vendor/symfony/console/Application.php(953): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#3 /home/pi/bin/carddav2fb/vendor/symfony/console/Application.php(248): Symfony\Component\Console\Application->doRunCommand(Object(Andig\RunCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#4 /home/pi/bin/carddav2fb/vendor/symfony/console/Application.php(148): Symfony\Component\Console\Application->doRun(Obj in /home/pi/bin/carddav2fb/src/functions.php on line 59
```

After a quick google research: It seems the installation steps (README.md) is missing a
```
sudo apt-get install php7.0-gd
``` 

=> I will provide a pull request for the installation steps in the README.md

### PR DESCRIPTION
**Pull request by derwok**

### CODE REVIEW

**Comment by andig** [comment]
Not sure about this PR. Why replace `file_put_contents` with the `ftp_*` functions?

**Comment by derwok** [comment]
OK, Sorry. I overlooked that... most times it's not good to submit a PR on master branch.
I'll clean up and close/remove this PR.

File: src/functions.php [hunk 1]
```diff
@@ -49,38 +51,57 @@ function download(Backend $backend, $substitutes, callable $callback=null): arra
...
+
+    // Prepare FTP connection
+    $ftpserver = $config['url'];
+    $ftpserver = str_replace("http://", "", $ftpserver);
+    $ftp_conn = ftp_connect($ftpserver);
+    if (!$ftp_conn) {
+        error_log("ERROR: Could not connect to ftp server ".$ftpserver." for image upload.");
+        return 0;
+    }
+    if (! ftp_login($ftp_conn, $config['user'], $config['password'])) {
```
Comment (andig) [inline src/functions.php]: remove " " after !

File: src/functions.php [hunk 2]
```diff
@@ -49,38 +51,57 @@ function download(Backend $backend, $substitutes, callable $callback=null): arra
...
+        error_log("ERROR: Could not log in ".$config['user']." to ftp server ".$ftpserver." for image upload.");
+        return 0;
+    } 
+    if (! ftp_chdir($ftp_conn, $config['fonpix'])){
```
Comment (andig) [inline src/functions.php]: dito

File: src/functions.php [hunk 3]
```diff
@@ -49,38 +51,57 @@ function download(Backend $backend, $substitutes, callable $callback=null): arra
...
+    $countUploadedImages = 0;
+    $countAllImages = 0;    
+
+    // Prepare FTP connection
+    $ftpserver = $config['url'];
+    $ftpserver = str_replace("http://", "", $ftpserver);
+    $ftp_conn = ftp_connect($ftpserver);
+    if (!$ftp_conn) {
+        error_log("ERROR: Could not connect to ftp server ".$ftpserver." for image upload.");
+        return 0;
```
Comment (andig) [inline src/functions.php]: remove return argument- the function doesn't return anything

File: src/functions.php [hunk 4]
```diff
@@ -49,38 +51,57 @@ function download(Backend $backend, $substitutes, callable $callback=null): arra
...
+    }
+    if (! ftp_login($ftp_conn, $config['user'], $config['password'])) {
+        error_log("ERROR: Could not log in ".$config['user']." to ftp server ".$ftpserver." for image upload.");
+        return 0;
```
Comment (andig) [inline src/functions.php]: dito

File: src/functions.php [hunk 5]
```diff
@@ -49,38 +51,57 @@ function download(Backend $backend, $substitutes, callable $callback=null): arra
...
+    } 
+    if (! ftp_chdir($ftp_conn, $config['fonpix'])){
+        error_log("ERROR: Could change to dir ".$config['fonpix']." on ftp server ".$ftpserver." for image upload.");
+        return 0;
```
Comment (andig) [inline src/functions.php]: dito

File: src/functions.php [hunk 6]
```diff
@@ -7,6 +7,8 @@
 use Andig\FritzBox\Converter;
 use Andig\FritzBox\Api;
 use \SimpleXMLElement;
+use Symfony\Component\Console\Output\ConsoleOutput;
```
Comment (andig) [inline src/functions.php]: please don't do that- it breaks reusability of the function and introduces tight coupling with symfony. If you eant progress, add a callback function instead.

File: src/functions.php [hunk 7]
```diff
@@ -49,38 +51,57 @@ function download(Backend $backend, $substitutes, callable $callback=null): arra
...
-                    }
-                    imagedestroy($imgFile);
+            if (preg_match("/JPEG/", strtoupper(substr($vcard->photoData, 0, 256)))) {     // Fritz!Box only accept jpg-files
+                $countAllImages++;
+                $remotefilename = sprintf('%1$s.jpg', $vcard->uid);
+                // We only upload if filesize differs. Non existing server file will have size -1
+                if (ftp_size($ftp_conn, $remotefilename) == strlen($vcard->rawPhoto)) {
+                    continue;
+                }
+                $memstream = fopen('php://memory', 'r+');     // we use a fast in-memory file stream
```
Comment (andig) [inline src/functions.php]: very good, thank you!

File: src/functions.php [hunk 8]
```diff
@@ -49,38 +51,57 @@ function download(Backend $backend, $substitutes, callable $callback=null): arra
...
+                fputs($memstream, $vcard->rawPhoto);
+                rewind($memstream);
+
+                // upload file
+                if (ftp_fput($ftp_conn, $remotefilename, $memstream, FTP_BINARY)){
+                    $countUploadedImages++;
+                } else {
+                    echo "Error uploading $remotefilename.\n";
```
Comment (andig) [inline src/functions.php]: rather error_log?

File: src/functions.php [hunk 9]
```diff
@@ -95,7 +116,11 @@ function dissolveGroups (array $vcards): array
 
     foreach ($vcards as $key => $vcard) {          // separate iCloud groups
         if (isset($vcard->xabsmember)) {
-            $groups[$vcard->fullname] = $vcard->xabsmember;
+            if(array_key_exists($vcard->fullname, $groups)) {
```
Comment (andig) [inline src/functions.php]: space after if

File: src/functions.php (hunk 6 above)
Comment (derwok) [reply src/functions.php]: Hmmm... sorry - unsure how to do that... can you point me to an example?

File: src/functions.php (hunk 2 above)
Comment (derwok) [reply src/functions.php]: OK

File: src/functions.php (hunk 6 above)
Comment (andig) [reply src/functions.php]: Check `setProgress` in backend.php:

            if (is_callable($this->callback)) {
                ($this->callback)();
            }

**Pull request closed without merging**